allennlp/modules/conditional_random_field.py

```diff
@@ -19,15 +63,27 @@ class ConditionalRandomField(torch.nn.Module):
     ----------
     num_tags : int, required
         The number of tags.
+    constraints : List[Tuple[int, int]], optional (default: None)
+        An optional list of allowed transitions (from_tag_id, to_tag_id).
+        These are applied to ``viterbi_tags()`` but do not affect ``forward()``.
     """
     def __init__(self,
-                 num_tags: int) -> None:
+                 num_tags: int,
+                 constraints: List[Tuple[int, int]] = None) -> None:
         super().__init__()
         self.num_tags = num_tags
 
         # transitions[i, j] is the logit for transitioning from state i to state j.
         self.transitions = torch.nn.Parameter(torch.Tensor(num_tags, num_tags))
 
+        # _constraint_mask indicates valid transitions (based on supplied constraints).
+        if constraints is None:
+            self._constraint_mask = None
+        else:
+            self._constraint_mask = torch.Tensor(num_tags, num_tags).fill_(0.)
```

Comment:
I don't think this is device agnostic, because when you call `.cuda` on a module, only it's `Parameters` are copied to the new device. It's going to be a little tricky to do this in this way though.

Reply:
I changed it to a `Parameter`, take a look and let me know if it looks right